c2rust/src/main.rs

```diff
@@ -1,56 +1,154 @@
+            });
         }
-    };
+        Ok(sub_commands)
+    }
+
+    /// Get all known [`SubCommand`]s.  These have no [`SubCommand::path`].
+    /// Even if the subcommand executables aren't there, we can still suggest them.
+    pub fn known() -> impl Iterator<Item = Self> {
+        ["transpile", "instrument", "pdg", "analyze"]
+            .into_iter()
+            .map(|name| Self {
+                path: None,
+                name: name.into(),
+            })
+    }
+
+    /// Get all known ([`Self::known`]) and actual, found ([`Self::find_all`]) subcommands,
+    /// putting the known ones first so that the found ones overwrite them and take precedence.
+    pub fn all() -> anyhow::Result<impl Iterator<Item = Self>> {
+        Ok(Self::known().chain(Self::find_all()?))
+    }
+
+    pub fn invoke<I, S>(&self, args: I) -> anyhow::Result<()>
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<OsStr>,
+    {
+        let path = self.path.as_ref().ok_or_else(|| {
+            anyhow!(
+                "known subcommand not found (probably not built): {}",
+                self.name
+            )
+        })?;
+        let status = Command::new(&path).args(args).status()?;
+        process::exit(status.code().unwrap_or(1));
+    }
 }
 
-fn invoke_subcommand<I, S>(subcommand: &str, args: I)
-where
-    I: IntoIterator<Item = S>,
-    S: AsRef<OsStr>,
-{
-    // Assumes the subcommand executable is in the same directory as this driver
-    // program.
-    let cmd_path = std::env::current_exe().expect("Cannot get current executable path");
-    let mut cmd_path = cmd_path.as_path().canonicalize().unwrap();
-    cmd_path.pop(); // remove current executable
-    cmd_path.push(format!("c2rust-{}", subcommand));
-    assert!(cmd_path.exists(), "{:?} is missing", cmd_path);
-    exit(
-        Command::new(cmd_path.into_os_string())
-            .args(args)
-            .status()
-            .expect("SubCommand failed to start")
-            .code()
-            .unwrap_or(-1),
-    );
+fn main() -> anyhow::Result<()> {
+    let sub_commands = SubCommand::all()?.collect::<Vec<_>>();
+    let sub_commands = sub_commands
+        .iter()
+        .map(|cmd| (cmd.name.as_ref(), cmd))
+        .collect::<HashMap<_, _>>();
+
+    // If the subcommand matches, don't use `clap` at all.
+    //
+    // I can't seem to get `clap` to pass through all arguments as is,
+    // like the ones with hyphens like `--metadata`,
+    // even though I've set [`Arg::allow_hyphen_values`].
+    // This is faster anyways.
+    // I also tried a single "subcommand" argument with [`Arg::possible_values`],
+    // but that had the same problem passing through all arguments as well.
+    //
+    // Furthermore, doing it this way correctly forwards `--help` through to the subcommand
+    // instead of `clap` intercepting it and displaying the top-level `--help`.
+    let mut args = env::args_os();
+    args.next(); // skip exe name
```

Comment:
You probably want to skip the second `args.next()` if the first one was `None`. Technically I think this is possible if you `fork()` and `exec()` manually, passing an empty `argv` array. Not realistic but good practice to handle that regardless. Probably can just skip straight to the clap arg parsing then.

Reply:
You're right.  I'll change it to use one `args.nth(1)`, which propagates the first `args.next()` properly.